### PR TITLE
refactor: web-extension e2e test setup with background wallet

### PR DESCRIPTION
**Pull request by mkazlauskas**
# Context

It is not obvious how to set up web extension with wallet running in the background

# Proposed Solution

Update `web-extension` package e2e tests to demonstrate setup using `SingleAddressWallet`, `LedgerKeyAgent` and blockfrost providers. Found and resolved a number of issues in the process. Key changes:
- `KeyAgent` has to run in UI script, because service workers don't support web hid.
  - Create new `AsyncKeyAgent` interface and update `keyAgent` dependency in `SingleAddressWallet` to it
- Fix a few issues in `web-extension` messaging:
  - Support 'lazy' subscriptions (before the other end exists)
  - Fix a few serialization issues (usage of toSerializableObj/fromSerializableObj)
- Multiple PouchDB storage fixes [839f705](https://github.com/input-output-hk/cardano-js-sdk/pull/263/commits/839f70553efd29488b2948a9e4e53992f3a7b454)...[419a419](https://github.com/input-output-hk/cardano-js-sdk/pull/263/commits/419a419635bebf653b530971c85de9b44b429f66)

For more details:
- see commit log
- see [web-extension setup diagram](https://drive.google.com/file/d/1kbv2moNRDAI1Not9G07EtZv5lMKZGgbB/view?usp=sharing)

# Important Changes Introduced

- fix(wallet): add communicationType arg to ledger getHidDeviceList [7ea7f8c](https://github.com/input-output-hk/cardano-js-sdk/pull/263/commits/7ea7f8c2f755f2bf7ce4e24c7aebea5ca8cfe955)
- refactor!: rm ObservableWallet.networkId (to be resolved via networkInfo$) [c76d917](https://github.com/input-output-hk/cardano-js-sdk/pull/263/commits/c76d9176ed7af7a8d036e7c12c48f1d9634656e4)
- fix(blockfrost): set correct testnet network magic [e4cca79](https://github.com/input-output-hk/cardano-js-sdk/pull/263/commits/e4cca79f4bfc8114f7fbf389aae9afa56f0d40f3)
- fix(wallet): optimize polling by not making some redundant requests [24df78d](https://github.com/input-output-hk/cardano-js-sdk/pull/263/commits/24df78d9dd3da0870a18c86901af88506f7a1551)

## Notes

- Didn't run `wallet` e2e tests after refactors, because they are currently broken in `master` (missing UtxoProvider in SingleAddressWallet setup)
- There are probably more possible optimizations for SingleAddressWallet polling. Would be great to test loading of asset metadata (tokens & nfts), but I don't have any in my wallet.

### CODE REVIEW

File: packages/wallet/src/persistence/pouchdbStores/pouchdbWalletStores.ts [hunk 1]
```diff
@@ -47,21 +55,24 @@ export const createPouchdbWalletStores = (walletName: string): WalletStores => {
...
+        /**
+         * Multiplied by 100k to distinguish between blockNo=1,index=0 and blockNo=0,index=1
+         * Assuming there can never be more >=100k transactions in a block
+         */
+        (blockNo * 100_000 + index).toString(),
+      logger
     ),
-    unspendableUtxo: new PouchdbUtxoStore(`${baseDbName}UnspendableUtxo`),
-    utxo: new PouchdbUtxoStore(`${baseDbName}Utxo`)
+    unspendableUtxo: new PouchdbUtxoStore(`${baseDbName}UnspendableUtxo`, undefined, logger),
```
Comment (rhyslbw) [inline packages/wallet/src/persistence/pouchdbStores/pouchdbWalletStores.ts:75]: Points to dependencies being rearranged to come before options

**Review by rhyslbw — COMMENTED** [review]
Looking really good @mkazlauskas. A few minor issues:

- The commit message of 839f70553efd29488b2948a9e4e53992f3a7b454 is truncated
- 61172499cbc122deae0acff7aa10610ac98959bf could be `tests` commit type
- Is the diagram being added superseding the one in `cip30`?

**Comment by mkazlauskas** [comment]
> The commit message of https://github.com/input-output-hk/cardano-js-sdk/commit/839f70553efd29488b2948a9e4e53992f3a7b454 is truncated

The `_` in the end of the message literally means the character. Is that what you mean?

Otherwise, the message has length of 73 chars. Should we aim for <=72 characters in message header for GitHub UI compatibility?

> https://github.com/input-output-hk/cardano-js-sdk/commit/61172499cbc122deae0acff7aa10610ac98959bf could be tests commit type

OK thanks. I wasn't sure about this one as it has changes in devDependencies. Looking at it now it has `"blake2b-no-wasm"` incorrectly listed under `dependencies` - will fix.

> Is the diagram being added superseding the one in cip30?

Yes. I'll remove it from `cip30` package and replace it with a link to the one under `web-extension`

File: packages/wallet/src/cip30.ts [hunk 2]
```diff
@@ -96,10 +102,12 @@ export const createWalletApi = (
     logger.warn('getCollateral is not implemented');
     return null;
   },
-  getNetworkId: async (): Promise<number> => {
+  getNetworkId: async (): Promise<Cardano.NetworkId> => {
     logger.debug('getting networkId');
     const wallet = await walletReady;
-    return wallet.getNetworkId();
+    const networkInfo = await firstValueFrom(wallet.networkInfo$);
+    // Review: consider adding networkInfo.network.id instead
```
[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/wallet/src/cip30.ts]: Updated the comment in [e44dd73](https://github.com/input-output-hk/cardano-js-sdk/pull/263/commits/e44dd7301f5ec3849faf29f45c0403c835993fc5)

File: packages/wallet/test/persistence/pouchdbStores.test.ts [hunk 3]
```diff
@@ -46,7 +47,8 @@ describe('pouchdbStores', () => {
   });
 
   describe('PouchdbCollectionStore', () => {
-    const createStore = () => new PouchdbCollectionStore<DocType>(dbName, ({ __unsupportedKey }) => __unsupportedKey);
+    const createStore = () =>
+      new PouchdbCollectionStore<DocType>(dbName, dummyLogger, ({ __unsupportedKey }) => __unsupportedKey);
```
Comment (rhyslbw) [inline packages/wallet/test/persistence/pouchdbStores.test.ts:51]: That's a bit awkward too. `undefined` is better than needing to import the dummyLogger. Maybe revert, and we can come back to this in a params logger consistency check?

Comment (mkazlauskas) [reply packages/wallet/test/persistence/pouchdbStores.test.ts]: See [cb36cbb](https://github.com/input-output-hk/cardano-js-sdk/pull/263/commits/cb36cbbbb8c19e41d034092ddc015cf8417a456d)

**Review by rhyslbw — APPROVED** [review]
Well done @mkazlauskas, that's a mighty effort.

File: packages/wallet/src/cip30.ts [hunk 4]
```diff
@@ -95,13 +102,17 @@ export const createWalletApi = (
     logger.warn('getCollateral is not implemented');
     return null;
   },
-  getNetworkId: async (): Promise<number> => {
+  getNetworkId: async (): Promise<Cardano.NetworkId> => {
     logger.debug('getting networkId');
-    return wallet.getNetworkId();
+    const wallet = await walletReady;
+    const networkInfo = await firstValueFrom(wallet.networkInfo$);
+    // TODO: use networkInfo.network.id instead of magicToNetworkId
```
Comment (mchappell) [inline packages/wallet/src/cip30.ts:109]: Does this need addressing, or is this techdebt for later?

File: packages/web-extension/e2e/extension/background/config.ts [hunk 5]
```diff
@@ -57,7 +59,12 @@ const blockfrostApi = [
 ].includes('blockfrost')
   ? (async () => {
       logger.debug('WalletProvider:blockfrost - Initializing');
-      const blockfrost = new BlockFrostAPI({ isTestnet, projectId: env.BLOCKFROST_API_KEY });
+      const blockfrost = new BlockFrostAPI({
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        adapter: axiosFetchAdapter as any, // type mismatch: adapter uses axios 0.26, while blockfrost-js uses 0.21
```
Comment (mchappell) [inline packages/web-extension/e2e/extension/background/config.ts:64]: 👍

**Review by mchappell — DISMISSED** [review]

**Review by rhyslbw — APPROVED** [review]

**Pull request merged**